File: Tree/560-symmetricBinaryTree.py
# Definition for a binary tree node.
from collections import deque
class TreeNode(object):
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
class Solution(object):
    def isSymmetric(self, root):
        """
        :type root: TreeNode
        :rtype: bool
        """
        q=deque([(root,root)])
        while q:
            a,b=q.popleft()
            if not a and not b:
                continue
            if not a or not b:
                return False
            if a.val!=b.val:
                return False
            q.append((a.left,b.right))
            q.append((a.right,b.left))
        
        return True

        # def f(a,b):
        #     if not a and not b:
        #         return True
        #     if not a or not b:
        #         return False
        #     return a.val==b.val and f(a.left,b.right) and f(a.right,b.left)
        
        # return f(root,root)
        
root=TreeNode(10);
one=TreeNode(1)
two=TreeNode(2)
two2=TreeNode(2)
three=TreeNode(3)
four=TreeNode(4)

root.left=two
root.right=two2

two.left=three
two.right=four

two2.left=four
two2.right=three

print(Solution().isSymmetric(root))

File: Tree/560-symmetricBinaryTree.cpp
#include<iostream>
#include<deque>

class TreeNode
{
    public:
        int m_v{};
        TreeNode* left{};
        TreeNode* right{};
        TreeNode(int v,TreeNode* l=nullptr, TreeNode* r=nullptr):m_v{v},left{l},right{r}{};
};


bool isSymmetric(const TreeNode* root)
{
    std::deque<std::pair<const TreeNode*,const TreeNode*>> q{};
    q.push_back(std::make_pair(root,root));

    while(!q.empty())
    {
        std::pair<const TreeNode*,const TreeNode*> cur=q.front();
        q.pop_front();
        const TreeNode *a=cur.first,*b=cur.second;
        if((!a)&&(!b))
            continue;
        if((!a)||(!b))
            return false;
        if(a->m_v!=b->m_v)
            return false;
        
        q.push_back(std::make_pair(a->left,b->right));
        q.push_back(std::make_pair(a->right,b->left));
        
    }
    return true;
}

bool dfoo(const TreeNode* a, const TreeNode* b)
{
    if(!a&&!b)
        return true;
    if(!a || !b)
        return false;
    return (a->m_v==b->m_v)&&(dfoo(a->left,b->right))&&(dfoo(a->right,b->left)); 
}


bool isSymmetric2(const TreeNode* root)
{
    return dfoo(root,root);
}


int main()
{
    TreeNode *root=new TreeNode(10);
    TreeNode *one=new TreeNode(1);
    TreeNode *two=new TreeNode(2);
    TreeNode *two2=new TreeNode(2);
    TreeNode *third=new TreeNode(3);
    TreeNode *four=new TreeNode(4);

    root->left=two;
    root->right=two2;
    two->left=third;
    two->right=four;
    two2->left=four;
    two2->right=third;

    std::cout<<isSymmetric(root);
    std::cout<<isSymmetric2(root);
}